app/controllers/settings_controller.rb

```diff
@@ -184,6 +184,39 @@ def pushover_callback
     return redirect_to "/settings"
   end
 
+  def mastodon_auth
+    instance_name = params[:mastodon_instance_name].to_s
+    return redirect_to Mastodon.oauth_auth_url(instance_name)
+  end
+
+  def mastodon_callback
+    if !params[:code].present?
+      flash[:error] = "Invalid OAuth state"
+      return redirect_to "/settings"
+    end
+
+    tok, username = Mastodon.token_and_user_from_code(params[:instance], params[:code])
```

Comment:
I think this creates duplicate `MastodonInstance`s when two users use the same `Instance`. You probably want `MastodonInstance.find_or_create_by`